### PR TITLE
Issue #2820056 by chr.fritsch: Be able to create and save a draft without changing the published article (workflow integration)

### PR DESCRIPTION
**Pull request by chrfritsch**
Make sure these boxes are checked before submitting your pull request - thank you!

- [x] All coding styles are fulfilled. ([How to check for cs issues?](https://github.com/BurdaMagazinOrg/thunder-dev-tools/blob/master/README.md#code-style-guidelines))
- [x] All tests are running locally. ([How to run the test?](https://github.com/BurdaMagazinOrg/thunder-distribution/blob/develop/docs/development.md#how-to-run-the-tests))
- [x] Necessary update hooks are provided.
- [x] User roles have correct access for new introduced permission.
- [x] Every thunder module has a README.md in its root. Follow [this guidelines](https://www.drupal.org/node/2181737), but we don't need every topic.
- [x] Code is covered with well-balanced amount of inline comments.

If you are really awesome, then your feature is covered by additional tests. Well done!

### CODE REVIEW

**Comment by dbosen** [comment]
Please add the state change from draft to archived

**Comment by dbosen** [comment]
The sidebar on the right does not show the current state, just "Published" and "Unpublished", it should show the current state instead.

**Comment by chrfritsch** [comment]
Things to do:
- Scheduler does not work correctly https://www.drupal.org/project/scheduler/issues/2798689#comment-12527743
- Default content get's not imported as published (might be related to some test failures) https://www.drupal.org/project/default_content/issues/2952963#comment-12533105

**Comment by alexpott** [comment]
I've discussed some of the issues we're facing with paragraphs and the other core committers. Catch in particular had some interesting points:
> my suggestion would be don’t integrate content_moderation with Thunder, use workspaces instead, any issues with workspaces you’ll also get with content_moderation but they’ll be feasible to fix without disgusting hacks.

The problems are that a single entity is not often a single entity. We struggling with paragraphs (https://www.drupal.org/project/paragraphs/issues/2807371) but there is also path aliases (https://www.drupal.org/project/drupal/issues/2856363) and taxonomy terms.

File: composer.json
Comment (dbosen) [inline composer.json]: This patch does not seem to be relevant for users. Patches that are only relevant for our tests should go into the apply_patches function in https://github.com/BurdaMagazinOrg/thunder-distribution/blob/8.x-2.x/scripts/travis/04-install-thunder.sh 
This should also take care of the failing drush make tests.

**Comment by IT-Cru** [comment]
BUG: Role content creator should not have permission to schedule publishing of content.

**Comment by dbosen** [comment]
Alignment of the selectbox is broken.

bildschirmfoto 2018-09-04 um 10 35 19

**Comment by dbosen** [comment]
Also, if only one option is available, remove the select box and just display the text - in the case of the screenshot in the other comment, the word "Draft should be written instead of the unusable select box.

**Comment by dbosen** [comment]
bildschirmfoto 2018-09-04 um 11 19 37

Change "Delete revision" to "Delete <state>" e.g. "Delete Draft".
Also change the "Published [Draft]" part to "[Draft] of published Article", or in general:
"[Moderation state] of <state> <content type>"

**Review by dbosen — CHANGES_REQUESTED** [review]
See change requests in comments

**Comment by dbosen** [comment]
bildschirmfoto 2018-09-04 um 12 00 13

Button should not be changed to "Save as" if content type is not moderated.

File: modules/thunder_article/src/Form/ThunderNodeForm.php
Comment (alexpott) [inline modules/thunder_article/src/Form/ThunderNodeForm.php]: A node form no?

File: composer.json
Comment (alexpott) [inline composer.json:132]: How come this change? Is there a version we're incompatible with?

File: config/install/user.role.restricted_editor.yml
Comment (alexpott) [inline config/install/user.role.restricted_editor.yml]: If the idea is that this user can't publish anything then maybe they shouldn't have this permission.

File: src/Plugin/Field/FieldWidget/ModerationStateWidget.php
Comment (alexpott) [inline src/Plugin/Field/FieldWidget/ModerationStateWidget.php:24]: I think it might work documenting why we are overriding this widget.

Comment (alexpott) [reply src/Plugin/Field/FieldWidget/ModerationStateWidget.php:24]: It appears to be a simply this widget to be a list of possible future states as thunder provides a different way to know the current state.

File: thunder.info.yml
Comment (alexpott) [inline thunder.info.yml]: I'd leave scheduler here as a separate line as it is going to end up installed. It feels like normal practice here to not rely on dependency install.

File: composer.json
Comment (alexpott) [inline composer.json]: This is not the latest version of the patch you've made. The latest version seems to have considerable changes. Also I'm concerned about the implementation of the unpublished state in this patch because at the moment it lists both draft and unpublished in the thunder use-case. Draft is not going to un-publish anything. It's going to create a new forward revision exactly the same as the published revision and the content will still be published on the site. I think that that is quite dangerous behaviour.

File: modules/thunder_article/src/Form/ThunderNodeForm.php
Comment (alexpott) [inline modules/thunder_article/src/Form/ThunderNodeForm.php]: I'm guessing that main purpose here is to allow users to start from the published revision again. Have we considered offering the ability to create a new draft based on the published revision?

File: modules/thunder_article/src/Form/ThunderNodeForm.php
Comment (alexpott) [inline modules/thunder_article/src/Form/ThunderNodeForm.php]: What's wrong with $entity->getRevisionUser()?

**Review by alexpott — COMMENTED** [review]
I've manually tested the functionality and it looks pretty good in terms of being quite clear for the user. Which is probably the trickiest part of this.

There is a gotcha with content moderation that we might want to make clear. Somehow. Path aliases are not versioned. So if you change the path alias on a draft and there is published content - the path alias changes for that too :(

File: composer.json
Comment (dbosen) [reply composer.json:132]: We have the policy, that patched modules have to be version restricted.

Comment (chrfritsch) [reply composer.json:132]: We have to pin the scheduler version until our patch gets committed and released. If we are not going to do this we get composer errors as soon as the patch landed.

File: modules/thunder_article/src/Form/ThunderNodeForm.php
Comment (dbosen) [reply modules/thunder_article/src/Form/ThunderNodeForm.php]: The purpose here is to be able to quickly dismiss the current draft, not necessarily to create a new one. But the use case of creating a new draft (that overrides an existing one) might be interesting as well, I will discuss this when we present the current state to our stake holders.

Comment (alexpott) [reply modules/thunder_article/src/Form/ThunderNodeForm.php]: But doesn't this get into misleading territory when there are multiple forward revisions? In a revisioning world the ideal thing is to not to do deletes.

File: modules/thunder_article/src/Form/ThunderNodeForm.php
Comment (chrfritsch) [reply modules/thunder_article/src/Form/ThunderNodeForm.php]: Didn't know that.

File: thunder.info.yml
Comment (chrfritsch) [reply thunder.info.yml]: Fixed

File: src/Plugin/Field/FieldWidget/ModerationStateWidget.php
Comment (chrfritsch) [reply src/Plugin/Field/FieldWidget/ModerationStateWidget.php:24]: Fixed

File: config/install/user.role.restricted_editor.yml
Comment (chrfritsch) [reply config/install/user.role.restricted_editor.yml]: True story. Removed the 'use editorial transition unpublished_published' permission

File: modules/thunder_article/src/Form/ThunderNodeForm.php
Comment (chrfritsch) [reply modules/thunder_article/src/Form/ThunderNodeForm.php]: Fixed

File: modules/thunder_article/src/Form/ThunderNodeForm.php
Comment (dbosen) [reply modules/thunder_article/src/Form/ThunderNodeForm.php]: Yes, that is a problem.

Comment (dbosen) [reply modules/thunder_article/src/Form/ThunderNodeForm.php]: The current idea is to change this to something like "Revert to last published version"

Comment (alexpott) [reply modules/thunder_article/src/Form/ThunderNodeForm.php]: 👍 That sounds like a good idea because:
* Non destructive
* Less deletes all in a row in the UI.

File: modules/thunder_article/src/Form/ThunderNodeForm.php
Comment (alexpott) [inline modules/thunder_article/src/Form/ThunderNodeForm.php:73]: Just thought about this some more. I think we should also copy better with multiple forward revisions by directing people to the list of revisions. This will also allow people to compare the revisions to see has has changed.

**Comment by klucic** [comment]
Hi All we just had a Review with Steffen and Mary and got the following Feedback

- ~~emphasize more that a Node is in "Draft Mode", as the current implementation is functionally working but is easy to oversee~~
- [x] deleting the current version should be replaced by the option to revert to the live version
- [x] remove redundand information of revision and publishing state (published of published, unpublished of unpublished)

File: src/Form/NodeRevisionRevertDefaultForm.php
Comment (alexpott) [inline src/Form/NodeRevisionRevertDefaultForm.php:16]: I think this functionality works really nicely.

Comment (alexpott) [reply src/Form/NodeRevisionRevertDefaultForm.php:16]: We do seem to be missing automated test coverage of this functionality though.

**Review by alexpott — COMMENTED** [review]
Manual testing shows everything works well. I do find the whole mixing of creating drafts and publishing and unpublishing odd for simple workflows like this. But that's part of having an extensible system.

**Review by dbosen — APPROVED** [review]

**Pull request merged**
